core/src/solve.jl

```diff
@@ -463,30 +586,39 @@ function water_balance!(du, u, p, t)::Nothing
         fractional_flow,
         flow_boundary,
         pump,
+        pid_control,
     ) = p
 
+    storage = collect(u.storage)
+    integral = collect(u.integral)
+
+    dstorage_old = copy(collect(du.storage))
```

Comment:
Why do you need `collect` here?

Regarding the `copy`, if possible we should avoid allocating arrays in our time loop. If we need it we should put it in one of the structs so we can allocate it once and reuse.